### ISSUE TITLE
Optimization recommendations for transport and serialization

### ISSUE THREAD

**Issue by codingmiao** [neo4j/neo4j#12551]
## Feature request template

**Is your feature request related to a problem? Please describe.**
When I query the data using Procedure, the Bolt-JSON transfer consumes too much time

**Describe the solution you'd like**
Add a high-performance protocol like ProtoBuffer

**Describe alternatives you've considered**
Procedure still returns Stream, but serialization and transport by protoBuffer 

**Additional context**
Here are some of my test results

Test1:

I did a Procedure that returned a list with a size of 20,000

server:
````
    public static class TopoResult {
        public final Long id;
        public final String devId;
        public final Boolean isConn;

        public TopoResult(Long id, String devId, Boolean isConn) {
            this.id = id;
            this.devId = devId;
            this.isConn = isConn;
        }
    }

    @Procedure("test.topo")
    @Description("topo测试")
    public Stream<TopoResult> topo(@Name("id") String id) {
        List<TopoResult> res = query(id);
        Stream<TopoResult> stream = res.stream();
        return stream;
    }
````
client:
```
        session.readTransaction(tx->{
            Result rs = tx.run("call test.topo($id)", Map.of(
                    "id", "123"
            ));
            rs.forEachRemaining((row) -> {
                row.get(0);
                row.get(1);
                row.get(2);
            });
            return tx;
        });
```
It took about 800ms to complete the query, of which serialization and transport took about 100ms

Test2:

I use ProtoBuffer to optimize the above program

server
```
    public static class TopoResultPb {
        public final byte[] b;

        public TopoResultPb(byte[] b) {
            this.b = b;
        }
    }

    @Procedure("test.topopb")
    @Description("topo测试pb")
    public Stream<TopoResultPb> topopb(@Name("psrurl") String psrurl) {
	List<TopoResult> list = query(id);
	byte[] bts = toPbfBytes(list) ;//Protobuffer to convert the List to byte[]
        	Stream<TopoResultPb> stream = Stream.of(bts);
	return stream;
    }
```

client
```
        session.readTransaction(tx -> {
            Result rs = tx.run("call test.topopb($id)", Map.of(
                    "id", "123"
            ));
            Record row = rs.next();
            byte[] bt = row.get(0).asByteArray();
            try {
                BoltProto.BoltPbArrPb arr = BoltProto.BoltPbArrPb.parseFrom(bt);
                for (BoltProto.BoltPb boltPb : arr.getArrList()) {
                    boltPb.getConn();
                    boltPb.getDevId();
                    boltPb.getId();
                }
            } catch (InvalidProtocolBufferException e) {
                e.printStackTrace();
            }
            return tx;
        });
```
It took about 700ms to complete the query, of which serialization and transport took about 10ms!

So, if Neo4j can use ProtoBuffer instead of JSON, that would be very efficient and nice!

**Comment by gjmwoods** [comment]
@codingmiao Thanks for the feature request. This is an interesting example, would you be able to share the complete setup so we can understand it fully and then discuss it within the team?

**Comment by codingmiao** [comment]
@gjmwoods 
I uploaded a demo here,
https://github.com/codingmiao/bolttest
You can run it by referring to Readme.md

**Comment by technige** [comment]
Hi @codingmiao

The Bolt protocol does not use JSON internally. It is designed using our on in-house syntax layer. When we designed this originally, we considered a wide range of options, including ProtoBuf, against a number of different criteria. As things stand now, while your measurements may indicate an improvement in pure on-wire transfer, this is rarely a bottleneck, and replacing our syntax layer would be a huge amount of work at this stage.

Closing as a wontfix.
